### PR TITLE
Tightening GRPCRoute validation, fixing gaps in HTTPRoute and Gateway webhook validation

### PR DESCRIPTION
**Pull request by robscott**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
This PR started with a single goal - address feedback from #1538 by filling in missing GRPCRoute webhook validation. As I started that process, I realized that we'd introduced some significant gaps between v1a2 and v1b1 validation among the resources that had graduated. This also updates that logic to share code where possible and provide a consistent experience.

**Does this PR introduce a user-facing change?**:
```release-note
- GRPCRoute: Regex validation for Method and Service has been tightened to match GRPC spec.
- GRPCRoute: Webhook validation of GRPCRoute has been expanded to closely match HTTPRoute validation.
- HTTPRoute and Gateway: Gaps between webhook validation for v1alpha2 and v1beta1 have been closed.
- HTTPRoute: Validating webhook now ensures that Exact and Prefix path match values can now only include valid path values per RFC-3986. (RegularExpression path matches are not affected by this change).
```

### CODE REVIEW

File: apis/v1alpha2/validation/grpcroute.go
Comment (howardjohn) [inline apis/v1alpha2/validation/grpcroute.go:102]: should this be a constraint on the CRD instead of the webhook? I think its preferred to have as much in the CRDs as possible?

File: apis/v1alpha2/validation/grpcroute.go
Comment (howardjohn) [inline apis/v1alpha2/validation/grpcroute.go:107]: This is already validated in the CRD:

```

// MethodMatchType specifies the semantics of how gRPC methods and services are compared.
// Valid MethodMatchType values are:
//
// * "Exact"
// * "RegularExpression"
//
// Exact paths MUST be syntactically valid:
//
// - Must not contain `/` character
//
// +kubebuilder:validation:Enum=Exact;PathPrefix;RegularExpression
// +kubebuilder:validation:Enum=Exact;RegularExpression
type GRPCMethodMatchType string
```

(note: it is listed 2x for some reason)

File: apis/v1alpha2/validation/gateway.go
Comment (shaneutt) [inline apis/v1alpha2/validation/gateway.go:35]: :+1:

File: apis/v1alpha2/validation/grpcroute.go
Comment (shaneutt) [inline apis/v1alpha2/validation/grpcroute.go:30]:
```suggestion
	// repeatableGRPCRouteFilters are filter types that are allowed to be
```

File: apis/v1alpha2/validation/grpcroute.go
Comment (shaneutt) [inline apis/v1alpha2/validation/grpcroute.go:37]: For future readers, it might be helpful to add comments to these explaining what, and maybe even some of the why.

File: apis/v1alpha2/validation/grpcroute.go
Comment (shaneutt) [inline apis/v1alpha2/validation/grpcroute.go:86]:
```suggestion
// validateGRPCMethod validates that no header name is matched more than once (case-insensitive),
// and that at least one of service or method was provided.
```

File: apis/v1alpha2/validation/grpcroute.go
Comment (shaneutt) [reply apis/v1alpha2/validation/grpcroute.go:102]: ++

Comment (youngnick) [reply apis/v1alpha2/validation/grpcroute.go:102]: The existing regex for method is `^[^\/]*$` (blocking only slash characters), so I guess we could add some stuff in there as well. But the idea here is to be belt-and-suspenders for this.

File: apis/v1alpha2/validation/grpcroute.go
Comment (youngnick) [reply apis/v1alpha2/validation/grpcroute.go:107]: I _think_ the first will be immediately overridden by the second, but it's incorrect anyway, so we should just remove it.

File: apis/v1alpha2/validation/grpcroute.go
Comment (youngnick) [reply apis/v1alpha2/validation/grpcroute.go:234]: Yes, seems like this one should also be `string(action.Name)` like the above.

**Comment by robscott** [comment]
Adding a hold since I want to wait for a couple reviews on this one.

/hold

File: apis/v1alpha2/validation/grpcroute.go
Comment (robscott) [reply apis/v1alpha2/validation/grpcroute.go]: This is actually copied verbatim from HTTPRoute. Although it looks very confusing, it's because the other modifiers are setting a header to a value so there's a name and value whereas here we're just removing headers so only the name is necessary. To make this easier to understand, I've named this var `name` instead of `action` for this section in both GRPCRoute and the original HTTPRoute validation.

File: apis/v1alpha2/validation/grpcroute.go
Comment (robscott) [reply apis/v1alpha2/validation/grpcroute.go:161]: Added, thanks!

File: apis/v1alpha2/validation/grpcroute.go
Comment (robscott) [reply apis/v1alpha2/validation/grpcroute.go]: Added, thanks!

File: apis/v1alpha2/validation/grpcroute.go
Comment (robscott) [reply apis/v1alpha2/validation/grpcroute.go]: Added, thanks!

File: apis/v1alpha2/validation/grpcroute.go
Comment (robscott) [reply apis/v1alpha2/validation/grpcroute.go]: Fixed here and in HTTPRoute, thanks!

File: apis/v1alpha2/validation/grpcroute.go
Comment (robscott) [reply apis/v1alpha2/validation/grpcroute.go]: I can't find that line doubled up, but let me know if it's still in main somewhere that I can't find. @howardjohn which specific part are you saying is validated in the CRD? That m.Method can't be nil?

File: apis/v1alpha2/validation/grpcroute.go
Comment (robscott) [reply apis/v1alpha2/validation/grpcroute.go]: I've been chatting with @gnossen about the best approach here and for now have tightened up CRD based validation since it appears that a safe regex is simple enough to write. Will hold off until I can get a clear reference to GRPC spec to confirm validations I've proposed here are accurate.

File: apis/v1alpha2/validation/grpcroute.go
Comment (youngnick) [reply apis/v1alpha2/validation/grpcroute.go]: Yeah, I can't find the double up either.

**Review by youngnick — COMMENTED** [review]
/lgtm

File: apis/v1alpha2/validation/grpcroute.go
Comment (gyohuangxin) [inline apis/v1alpha2/validation/grpcroute.go:72]: There are two different methods to handle `return errs` in webhook validation now.
1. If the condition matches, return errors immediately to avoid returning too many errors :
https://github.com/kubernetes-sigs/gateway-api/blob/df03bd58002fcf01ee3ffeba97edc1f0c67fb386/apis/v1beta1/validation/common.go#L52-L54
2. Return errors at the end of the function like your code.

I think we should unify the ways to return errors to users, returning all errors or returning the first error. What do you think?

File: apis/v1alpha2/validation/grpcroute.go
Comment (gyohuangxin) [inline apis/v1alpha2/validation/grpcroute.go:98]: Please remove the blank line.

File: apis/v1alpha2/grpcroute_types.go
Comment (gnossen) [inline apis/v1alpha2/grpcroute_types.go:325]: Pulling from the spec at protobuf.com. method is an `Identifier`. Translating from the BNF, I think it is correct.

Service is a [`TypeName`](https://protobuf.com/docs/language-spec#type-references), so I think there are two minor details off with this service regex though.

- It's underscores that are allowed, not dashes.
- We allow an optional leading dot

So I think the full regex should be the following ([regex101](https://regex101.com/r/bXTWuy/1)):

`^(?i)\.?[a-z_][a-z_0-9]+(\.[a-z_][a-z_0-9]+)*$`

This matches:
- `foo.bar.Baz`
- `.foo.bar.Baz`
- `.foo.bar.Bar_Baz`
- `foo`

Comment (robscott) [reply apis/v1alpha2/grpcroute_types.go]: Thanks @gnossen! Updated to match this, PTAL

File: apis/v1alpha2/validation/grpcroute.go
Comment (robscott) [reply apis/v1alpha2/validation/grpcroute.go:72]: If needed we can file a follow up for this since it doesn't feel directly tied to this PR. In general I think we should return as many errors as we reasonably can so users don't have take multiple iterations to fix a set of problems. I think that largely matches what we're already doing, but your specific example may be an example where it makes sense to return both errors.

File: apis/v1alpha2/validation/grpcroute.go
Comment (robscott) [reply apis/v1alpha2/validation/grpcroute.go:98]: I think this may just be a preference thing but I actually prefer a bit more whitespace like this. In this case, this func is copied almost verbatim from HTTPRoute and maintains the same spacing as the original function:

https://github.com/kubernetes-sigs/gateway-api/blob/df03bd58002fcf01ee3ffeba97edc1f0c67fb386/apis/v1beta1/validation/httproute.go#L180-L196

If you feel strongly about where we should have new lines in our functions it may be worth a follow up issue to discuss more.

**Review by gnossen — APPROVED** [review]

**Comment by robscott** [comment]
Still need one more LGTM, but removing hold since I think we have consensus on the changes here.

/hold cancel

**Review by shaneutt — APPROVED** [review]
/lgtm

**Comment by jackstine** [comment]
I had a question about the validation [here](https://github.com/kubernetes-sigs/gateway-api/commit/cd3a10073bc8243fbc49dd242acdac376323cd53#commitcomment-94626827)

**Pull request merged**
